src/Chapter3.hs

```diff
@@ -1007,6 +1112,61 @@ contestants, and write a function that decides the outcome of a fight!
 -}
 
 
+data Act = Heal | Defend | Attack | RunAway deriving (Show)
+
+data K = K
+  { healthK :: Int
+  , defenceK :: Int
+  , attackK :: Int
+  , nameK :: String
+  }
+
+data M = M
+  { healthM :: Int
+  , attackM :: Int
+  , nameM :: String
+  }
+
+class Fighter a where
+  name :: a -> String
+  moves :: a -> [Act]
```

Comment:
With this design it's not possible for different values of the `K` type to have different lists of actions.

Also, only `Knight` can cast spells, but I see that you have a single data type `Act` allowing monsters to have this type of action as well. Ideally, it's better to model data types in such way so that illegal states are impossible to happen